File: apps/api/src/rx_jev_api/routers/ask.py
"""What one label says about a reader's own question, asked live.

The question becomes part of the Jev instructions (`judge.build_custom_request`); the five
answer categories and the verbatim-quote rule stay the same as for catalog questions. Custom
answers are never stored and never reviewed, so every ask calls Jev, which is why asks are
limited in length and rate per client.
"""

import math
from typing import Annotated, Literal

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, StringConstraints

from rx_jev_api.deps import AskLimiterDep, JudgeDep, OpenFdaDep, RxNormDep, SettingsDep
from rx_jev_api.judge import CUSTOM, build_custom_request, question_key
from rx_jev_api.routers.answers import (
    AnswerStatus,
    EvidenceView,
    LabelInfo,
    StanceView,
    evidence_view,
    is_confident,
    label_info,
)
from rx_jev_api.routers.labels import RxCuiPath, canonical_labels

router = APIRouter(prefix="/api/labels", tags=["ask"])

# A question, not a document: long enough for a sentence, short enough to stay a topic.
MIN_QUESTION_CHARS = 3
MAX_QUESTION_CHARS = 200


class AskRequest(BaseModel):
    # Which of the drug's canonical labels to read.
    set_id: str = Field(min_length=1, max_length=64)
    question: Annotated[
        str,
        StringConstraints(
            strip_whitespace=True, min_length=MIN_QUESTION_CHARS, max_length=MAX_QUESTION_CHARS
        ),
    ]


class CustomAnswer(BaseModel):
    question: str
    status: AnswerStatus
    # Custom answers are never checked by a pharmacist.
    reviewed: Literal[False] = False
    # Same rule as catalog answers: both confidences at the threshold, and a quote.
    confident: bool
    stance: StanceView | None
    evidence: EvidenceView | None


class AskResponse(BaseModel):
    rxcui: str
    label: LabelInfo
    # None when the question was skipped, so Jev was never asked.
    model_version: str | None
    answer: CustomAnswer


@router.post("/{rxcui}/ask")
def ask(
    rxcui: RxCuiPath,
    body: AskRequest,
    request: Request,
    rxnorm: RxNormDep,
    openfda: OpenFdaDep,
    judge: JudgeDep,
    settings: SettingsDep,
    limiter: AskLimiterDep,
) -> AskResponse:
    # A spent limit is refused before any upstream call, but an ask is counted only once
    # its label is found, so a request that fails validation or finds no label costs nothing.
    client = request.client.host if request.client else "unknown"
    _refuse_if_waiting(limiter.wait(client))

    _, labels = canonical_labels(rxcui, rxnorm, openfda)
    label = next((lb for lb in labels if lb.set_id == body.set_id), None)
    if label is None:
        raise HTTPException(
            status_code=404, detail=f"Label {body.set_id} is not a current label of this drug."
        )
    _refuse_if_waiting(limiter.hit(client))

    judge_request = build_custom_request(label, body.question)
    skipped = judge_request.skipped.get(CUSTOM)
    if skipped is not None:
        answer = CustomAnswer(
            question=body.question,
            status=skipped,
            confident=False,
            stance=None,
            evidence=None,
        )
        return AskResponse(rxcui=rxcui, label=label_info(label), model_version=None, answer=answer)

    result = judge.judge(judge_request)
    stance = result.distributions[question_key(CUSTOM, "stance")]
    evidence = result.distributions[question_key(CUSTOM, "evidence")]
    answer = CustomAnswer(
        question=body.question,
        status="judged",
        confident=is_confident(stance, evidence, settings.display_min_confidence),
        stance=StanceView.model_validate(stance.model_dump()),
        evidence=evidence_view(judge_request.asked[CUSTOM], evidence),
    )
    return AskResponse(
        rxcui=rxcui, label=label_info(label), model_version=result.model_version, answer=answer
    )


def _refuse_if_waiting(wait: float | None) -> None:
    if wait is None:
        return
    seconds = max(1, math.ceil(wait))
    raise HTTPException(
        status_code=429,
        detail=f"Too many questions. Try again in {_duration(seconds)}.",
        headers={"Retry-After": str(seconds)},
    )


def _duration(seconds: int) -> str:
    if seconds < 60:
        return f"{seconds} seconds" if seconds != 1 else "1 second"
    minutes = math.ceil(seconds / 60)
    if minutes < 60:
        return f"{minutes} minutes" if minutes != 1 else "1 minute"
    hours = math.ceil(minutes / 60)
    return f"{hours} hours" if hours != 1 else "1 hour"

File: apps/worker/src/routes/ask.ts
import { type AskResponse, AskRequestSchema, type CustomAnswer } from '@rx-jev/contract'
import { Hono } from 'hono'
import type { AppEnv } from '../app.ts'
import { buildCustomRequest, CUSTOM, questionKey } from '../judge.ts'
import { canonicalLabels, type Lookup } from '../lookup.ts'
import { ProblemError, parseOr422 } from '../problems.ts'
import { candidateTrace, evidenceView, isConfident, stanceView } from '../views.ts'
import { labelInfo, RxCuiParams } from './labels.ts'

// What one label says about a reader's own question, asked live.
//
// The question becomes part of the Jev instructions (`buildCustomRequest`); the five answer
// categories and the verbatim-quote rule stay the same as for catalog questions. Custom
// answers are never stored and never reviewed, so every ask calls Jev, which is why asks
// are limited in length and rate per client.

async function readBody(request: Request): Promise<unknown> {
  try {
    return await request.json()
  } catch {
    throw new ProblemError(422, 'Invalid request: body')
  }
}

function duration(seconds: number): string {
  const unit = (n: number, name: string) => (n === 1 ? `1 ${name}` : `${n} ${name}s`)
  if (seconds < 60) return unit(seconds, 'second')
  const minutes = Math.ceil(seconds / 60)
  if (minutes < 60) return unit(minutes, 'minute')
  return unit(Math.ceil(minutes / 60), 'hour')
}

function refuse(wait: number): never {
  const seconds = Math.max(1, Math.ceil(wait))
  throw new ProblemError(429, `Too many questions. Try again in ${duration(seconds)}.`, {
    'Retry-After': String(seconds),
  })
}

export const askRoutes = new Hono<AppEnv>().post('/:rxcui/ask', async (c) => {
  const { rxcui } = parseOr422(RxCuiParams, c.req.param(), 'path')
  const body = parseOr422(AskRequestSchema, await readBody(c.req.raw), 'body')
  const { rxnorm, openfda, judge, config, askLimiter } = c.var.services

  // The ask takes its slot before any upstream call, so a burst cannot all reach RxNorm and
  // openFDA at once. An ask whose drug or label is not found gives back its own slot.
  const client = c.req.header('CF-Connecting-IP') ?? 'unknown'
  const slot = await askLimiter.acquire(client)
  if (typeof slot === 'number') refuse(slot)

  let lookup: Lookup
  let label: Lookup['labels'][number]
  try {
    lookup = await canonicalLabels(rxcui, rxnorm, openfda)
    const found = lookup.labels.find((l) => l.set_id === body.set_id)
    if (found === undefined) {
      throw new ProblemError(404, `Label ${body.set_id} is not a current label of this drug.`)
    }
    label = found
  } catch (error) {
    if (error instanceof ProblemError && error.status === 404) await askLimiter.release(slot)
    throw error
  }

  const request = await buildCustomRequest(label, body.question)
  const skipped = request.skipped[CUSTOM]
  if (skipped !== undefined) {
    const answer: CustomAnswer = {
      question: body.question,
      status: skipped,
      reviewed: false,
      confident: false,
      stance: null,
      evidence: null,
      candidates: 0,
      sections: [],
    }
    return c.json<AskResponse>({
      rxcui,
      label: labelInfo(label),
      model_version: null,
      input_tokens: null,
      output_tokens: null,
      latency_ms: null,
      answer,
      sources: lookup.sources,
    })
  }

  const result = await judge.judge(request)
  const stance = result.distributions[questionKey(CUSTOM, 'stance')]
  const evidence = result.distributions[questionKey(CUSTOM, 'evidence')]
  const candidates = request.asked[CUSTOM] ?? []
  if (!stance || !evidence) throw new Error('Jev returned no custom answer.')
  return c.json<AskResponse>({
    rxcui,
    label: labelInfo(label),
    model_version: result.model_version,
    input_tokens: result.input_tokens,
    output_tokens: result.output_tokens,
    latency_ms: result.latency_ms,
    answer: {
      question: body.question,
      status: 'judged',
      reviewed: false,
      confident: isConfident(stance, evidence, config.displayMinConfidence),
      stance: stanceView(stance),
      evidence: evidenceView(candidates, evidence),
      ...candidateTrace(candidates),
    },
    sources: lookup.sources,
  })
})
